llmfoundry/data/finetuning/dataloader.py

```diff
@@ -123,105 +124,110 @@ def build_finetuning_dataloader(cfg: DictConfig,
-                        +
-                        f'is {minimum_dataset_size} because you are running on {world_size} gpus and '
-                        +
-                        f'your per device batch size is {dataloader_batch_size}. Please increase the number '
-                        +
-                        f'of samples in your dataset to at least {minimum_dataset_size}.'
-                    )
-
-        assert dataset is not None
-        dl = DataLoader(
-            dataset,
-            collate_fn=collate_fn,
-            batch_size=dataloader_batch_size,
-            drop_last=cfg.drop_last,
-            sampler=dist.get_sampler(dataset,
-                                     drop_last=cfg.drop_last,
-                                     shuffle=cfg.dataset.shuffle),
-            num_workers=cfg.num_workers,
-            pin_memory=cfg.get('pin_memory', True),
-            prefetch_factor=cfg.get('prefetch_factor', 2),
-            persistent_workers=cfg.get('persistent_workers', True),
-            timeout=cfg.get('timeout', 0),
+            hf_name = _download_remote_dataset(hf_name, split)
+        elif cfg.dataset.get('safe_load') is True:
+            # Download dataset from huggingface hub with restrictions.
+            hf_kwargs = cfg.dataset.get('hf_kwargs', None)
+            token = hf_kwargs.get('token', None)
+            hf_name = _safe_download_hf_dataset(hf_name, token)
+        
+        # Build HF dataset
+        dataset = dataset_constructor.build_from_hf(
+            dataset_name_or_path=hf_name,
+            split=split,
+            max_seq_len=cfg.dataset.max_seq_len,
+            proto_preprocessing_fn=cfg.dataset.preprocessing_fn,
+            tokenizer=tokenizer,
+            hf_kwargs=cfg.dataset.get('hf_kwargs', {})
         )
+        sampler = dist.get_sampler(dataset,
+                    drop_last=cfg.drop_last,
+                    shuffle=cfg.dataset.shuffle)
+
+    collate_fn, dataloader_batch_size = _build_collate_fn(
+        cfg, tokenizer, device_batch_size)
+
+    if cfg.drop_last:
+        world_size = dist.get_world_size()
+        minimum_dataset_size = world_size * dataloader_batch_size
+        if hasattr(dataset, '__len__'):
+            full_dataset_size = len(dataset)
+            if full_dataset_size < minimum_dataset_size:
+                raise ValueError(
+                    f'Your dataset (name={cfg.dataset.hf_name}, split={cfg.dataset.split}) '
+                    +
+                    f'has {full_dataset_size} samples, but your minimum batch size '
+                    +
+                    f'is {minimum_dataset_size} because you are running on {world_size} gpus and '
+                    +
+                    f'your per device batch size is {dataloader_batch_size}. Please increase the number '
+                    +
+                    f'of samples in your dataset to at least {minimum_dataset_size}.'
+                )
+    
```

Comment:
@dakinggg Is this incorrect to have this check for both streaming and hf datasets? Originally, this check was only running for hf datasets.

Reply:
i didn't think too hard about which checks could be applied to streaming datasets safely. totally possible this should be applied in both places.

Reply:
@karan6181 Do you have any thoughts here? Should we be doing this check for streaming dataset as well? I haven't considered data duplication that might happen automatically in streaming, which might make this check incorrect

Reply:
one thing is that `len(dataset)` means something different for a streaming dataset (its per device) than a map style hf dataset (full dataset).